src/upload-reducer.tsx

```diff
@@ -2,39 +2,57 @@ export const SET_ERROR = 'SET_ERROR';
 export const START_UPLOADING = 'START_UPLOADING';
 export const SET_UPLOAD_PROGRESS = 'SET_UPLOAD_PROGRESS';
 export const FINISH_UPLOADING = 'FINISH_UPLOADING';
+export const CLEAR_STATE = 'CLEAR_STATE';
 
 export type UploadState = {
-  loading?: boolean;
-  progress?: number;
-  error?: string;
-  done?: boolean;
+  loading: boolean;
+  progress: number;
+  error: string | null;
+  done: boolean;
   response?: any;
 };
 
+export const initialState: UploadState = {
+  loading: false,
+  progress: 0,
+  error: null,
+  done: false,
+  response: null
+};
+
 export type Action =
   | { type: 'START_UPLOADING' }
   | { type: 'SET_UPLOAD_PROGRESS'; payload: number }
   | { type: 'SET_ERROR'; payload: string }
-  | { type: 'FINISH_UPLOADING'; payload: any };
+  | { type: 'FINISH_UPLOADING'; payload: any }
+  | { type: 'CLEAR_STATE'; };
 
 export type dispatchType = (action: Action) => void;
+
 // The possible state changes that take place during a file upload
-export function reducer(state: UploadState, action: Action) {
+export function reducer(state: UploadState, action: Action): UploadState {
   switch (action.type) {
     case START_UPLOADING:
-      return { loading: true };
+      return { ...state, loading: true };
```

Comment:
i think this should spread initialState, just in case it starts an upload without someone calling reset

Reply:
If I'm not mistaken, `initialState` will always be in place as it's passed in [here](https://github.com/zackify/react-use-upload/pull/22/files#diff-0ede280470af9fe188ee23deb087239fR54) as an initial value.

Reply:
but what happens if you call start_uploading, then it completes, so the other actions are called.

and THEN you start an upload again, it would be spreading the finished upload state, instead of being reset (in case the user did not reset before uploading twice)

i havent used the library in a bit, so let me know if im thinking incorrectly

Reply:
Ah, that's a good point.  My thinking on this was biased as I'm using this library as a "back-end" for a UI component that forces the user to clear the state before a second upload is initiated.  But you're correct: the library itself should not rely on `reset` being called before a second upload occurs.  I'll make the change now.

Excellent catch!